File: crates/rsi/core/tests/service_host_cli/tui/external.rs
use super::*;

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn external_tui_permissions_detach_cancel_and_reap_with_or_without_native_model() {
    for configured in [true, false] {
        let fixture = CliFixture::new("http://127.0.0.1:1");
        if !configured {
            std::fs::remove_file(fixture.temporary.path().join("config/rsi/settings.json"))
                .unwrap();
        }
        configure(&fixture);
        let mut terminal = TerminalClient::start(&fixture, &[]);
        terminal.capture_name = format!("external-model-{configured}");
        let native_hint = if configured {
            "Ctrl+J adds a line"
        } else {
            "No default model"
        };
        terminal.until(native_hint).await;
        terminal.send(b"/external\r");
        terminal.select_menu("Start fixture-external").await;
        terminal.until("Connected · Ready").await;
        terminal.send(b"permission\r");
        terminal.until("1 pending permissions").await;
        terminal.send(b"\x1b");
        terminal.until(native_hint).await;
        terminal.send(b"/attention\r");
        terminal.select_menu("Review permission").await;
        terminal.until("Read fixture · reject_always").await;
        terminal.select_menu("Read fixture · allow_always").await;
        terminal.until("fixture-permission-answered").await;
        terminal.until("Completed").await;
        terminal.send(b"permission\r");
        terminal.until("1 pending permissions").await;
        terminal.send(b"\x10");
        terminal.select_menu("Read fixture · reject_once").await;
        terminal.until("Completed").await;
        terminal.send(b"EXTERNAL_DRAFT_KEEP\x1b");
        terminal.until(native_hint).await;
        assert_eq!(
            std::fs::read_to_string(fixture.workspace.join("new-count")).unwrap(),
            "new\n"
        );
        let pid = std::fs::read_to_string(fixture.workspace.join("peer-pids")).unwrap();
        assert!(std::path::Path::new("/proc").join(pid.trim()).exists());
        terminal.send(b"/external\r");
        terminal.select_menu("fixture-external").await;
        terminal.until("EXTERNAL_DRAFT_KEEP").await;
        terminal.resize(PtySize {
            rows: 28,
            cols: 48,
            pixel_width: 0,
            pixel_height: 0,
        });
        terminal
            .until_screen("narrow external composer", |screen| {
                screen.contains("EXTERNAL_DRAFT_KEEP") && screen.contains("Enter select · Esc back")
            })
            .await;
        terminal.resize(PtySize {
            rows: 30,
            cols: 110,
            pixel_width: 0,
            pixel_height: 0,
        });
        terminal.until("Ctrl+P permissions and controls").await;
        terminal.send(b"\x15wait\r");
        terminal.until("Running").await;
        terminal.send(b"\x03");
        terminal.until("Cancelled").await;
        terminal.send(b"\x10");
        terminal.select_menu("Close peer").await;
        terminal.until("Closed").await;
        assert!(!std::path::Path::new("/proc").join(pid.trim()).exists());
        terminal.send(b"\x1b");
        terminal.until(native_hint).await;
        terminal.send(b"\x04");
        terminal.finish().await;
    }
}

fn configure(fixture: &CliFixture) {
    let profile = fixture
        .temporary
        .path()
        .join("config/rsi/host-profiles/fixture/host.profile.toml");
    let script = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("../../../fixtures/rsi/acp/agent.py")
        .canonicalize()
        .unwrap();
    let secret = serde_json::json!({"kind":"literal","value":"private-fixture-secret"});
    let config = serde_json::json!({"directory":fixture.temporary.path().join("state/rsi/acp"),"endpoints":[{"id":"fixture-external","enabled":true,"cwd":fixture.workspace,"sandbox":"danger-full-access","launch":{"program":"/usr/bin/python3","arguments":["-u",script,fixture.workspace,"normal"],"environment":{"FIXTURE_SECRET":secret}},"mcp_servers":[{"name":"private","launch":{"program":"/usr/bin/python3","environment":{"MCP_SECRET":secret}}}]}]});
    let mut document: toml::Value =
        toml::from_str(&std::fs::read_to_string(&profile).unwrap()).unwrap();
    document["steps"].as_array_mut().unwrap().push(
        toml::Value::try_from(
            serde_json::json!({"kind":"patch","target":"rsi-acp","config":config}),
        )
        .unwrap(),
    );
    std::fs::write(profile, toml::to_string(&document).unwrap()).unwrap();
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn delegated_tool_card_opens_the_existing_host_conversation() {
    let (endpoint, state, provider) = gated_provider("external_agent").await;
    *state.arguments.lock().unwrap() =
        Some(serde_json::json!({"operation":"start","endpoint":"fixture-external"}));
    let fixture = CliFixture::new(&endpoint);
    configure(&fixture);
    let mut terminal = TerminalClient::start(&fixture, &["--session-id", "delegation-card"]);
    terminal.capture_name = "delegation-card".into();
    terminal.until("Ctrl+J adds a line").await;
    terminal.send(b"Delegate this task\r");
    tokio::time::timeout(Duration::from_secs(20), state.requested.notified())
        .await
        .unwrap();
    state.release.notify_one();
    terminal.until("hello from daemon").await;
    terminal.send(b"\t\x10");
    terminal.select_menu("Raw sources / full output").await;
    terminal.select_menu("Open external conversation").await;
    terminal.until("Connected · Ready").await;
    assert_eq!(
        std::fs::read_to_string(fixture.workspace.join("new-count")).unwrap(),
        "new\n"
    );
    assert_eq!(state.requests.lock().unwrap().len(), 2);
    terminal.capture();
    terminal.send(b"work\r");
    terminal.until("fixture-output").await;
    terminal.send(b"\x10");
    terminal.select_menu("Close peer").await;
    terminal.until("Closed").await;
    let pid = std::fs::read_to_string(fixture.workspace.join("peer-pids")).unwrap();
    assert!(!std::path::Path::new("/proc").join(pid.trim()).exists());
    terminal.send(b"\x1b");
    terminal.until("Enter actions").await;
    terminal.send(b"\x1b");
    terminal.until("End follows output").await;
    terminal.send(b"\x04");
    terminal.finish().await;
    provider.abort();
}

File: fixtures/rsi/desktop-product/external.py
"""External ACP controls through actual WebKitGTK elements and the native bridge."""
import json
import os
from pathlib import Path
import shutil


def configure(profile, workspace, state):
    node = shutil.which('node')
    assert node, 'Independent ACP SDK fixture requires an explicit Node executable'
    agent = Path(__file__).resolve().parents[1] / 'acp/agent.mjs'
    q = json.dumps
    profile.write_text(profile.read_text().replace('steps = []', ''))
    with profile.open('a') as out:
        out.write(f'''\n[[steps]]
kind='patch'
target='rsi-acp'
[steps.config]
directory={q(str(state / 'acp'))}
[[steps.config.endpoints]]
id='sdk-agent'
enabled=true
cwd={q(str(workspace))}
sandbox='danger-full-access'
[steps.config.endpoints.launch]
program={q(node)}
arguments=[{q(str(agent))},{q(str(workspace))}]
[steps.config.endpoints.launch.environment]
FIXTURE_SECRET={{kind='literal',value='private-fixture-secret'}}
[[steps.config.endpoints.mcp_servers]]
name='fixture-mcp'
[steps.config.endpoints.mcp_servers.launch]
program='/usr/bin/python3'
[steps.config.endpoints.mcp_servers.launch.environment]
MCP_SECRET={{kind='literal',value='private-fixture-secret'}}
''')


def verify(script, button, fill, until, screenshot, workspace, report):
    button('Start sdk-agent')
    until(lambda: script('return !!document.querySelector(".external-composer")'))
    fill('[aria-label="External message"]', 'permission')
    button('Send ↗')
    until(lambda: script('return document.querySelectorAll(".external-permission button").length===4'))
    until(lambda: script('return [...document.querySelectorAll(".attention-navigation button")].some(e=>e.textContent==="Review permission 1")'))
    button('Review permission 1')
    until(lambda: script('return document.activeElement?.classList.contains("external-permission")'))
    screenshot('external-permission.png')
    button('Always · allow always')
    until(lambda: script('return document.querySelector(".external-status")?.textContent.includes("completed")'))
    until(lambda: script('return document.querySelector(".external-transcript")?.textContent.includes("SDK_AGENT_VERIFIED")'))
    assert script('return window.externalExecuted===undefined')
    screenshot('external-conversation.png')
    fill('[aria-label="External message"]', 'reject-always')
    button('Send ↗')
    until(lambda: script('return document.querySelectorAll(".external-permission button").length===4'))
    button('Never · reject always')
    until(lambda: script('return document.querySelector(".external-status")?.textContent.includes("completed")'))
    fill('[aria-label="External message"]', 'wait')
    button('Send ↗')
    until(lambda: script('return document.querySelector(".external-status")?.textContent.includes("running")'))
    button('Cancel prompt')
    until(lambda: script('return document.querySelector(".external-status")?.textContent.includes("cancelled")'))
    button('Close peer')
    until(lambda: script('return document.querySelector(".external-status")?.textContent.includes("Closed")'))
    button('Reload remote history')
    until(lambda: script('return [...document.querySelectorAll(".external-record pre")].some(e=>e.textContent==="1199")'),45)
    assert script('return document.querySelectorAll(".external-record").length') == 128
    screenshot('external-replay.png')
    button('Close peer')
    until(lambda: script('return document.querySelector(".external-status")?.textContent.includes("Closed")'))
    pids = (workspace / 'peer-pids').read_text().splitlines()
    for pid in pids:
        try:
            os.kill(int(pid), 0)
        except ProcessLookupError:
            continue
        raise AssertionError(f'ACP peer {pid} was not reaped')
    (report / 'external.json').write_text(json.dumps({'sdk':'1.4.0','permission_options':4,'permission_requested_again':True,'cancel':True,'replay':1200,'retained':128,'peers_reaped':len(pids)},indent=2))


def provider_reply(body):
    messages = body.get('messages', [])
    index = next((i for i in range(len(messages)-1, -1, -1) if messages[i]['role']=='user'), -1)
    if index < 0 or 'start an external delegation' not in str(messages[index].get('content', '')):
        return None
    if any(m['role']=='tool' for m in messages[index+1:]):
        return None
    return {'tool_calls':[{'index':0,'id':'desktop-delegate','type':'function','function':{'name':'external_agent','arguments':json.dumps({'operation':'start','endpoint':'sdk-agent'})}}]}


def delegation(script, button, fill, until, screenshot, workspace, report):
    native_session = script('return document.querySelector(".pane-session").title')
    before = (workspace / 'new-count').read_text()
    fill('textarea[aria-label="Main message"]', 'Please start an external delegation')
    button('Send ↗')
    until(lambda: script('return [...document.querySelectorAll(".transcript button")].some(b=>b.textContent==="Open external conversation")'))
    until(lambda: script('return document.querySelector(".pane-status")?.textContent.includes("Completed")'))
    screenshot('delegation-card.png')
    button('Open external conversation')
    until(lambda: script('return document.querySelector(".external-status")?.textContent.includes("ready")'))
    assert (workspace / 'new-count').read_text() == before + 'new\n'
    fill('[aria-label="External message"]', 'work')
    button('Send ↗')
    until(lambda: script('return document.querySelectorAll(".external-permission button").length===4'))
    button('Always · allow always')
    until(lambda: script('return document.querySelector(".external-status")?.textContent.includes("completed")'))
    screenshot('delegation-conversation.png')
    button('Close peer')
    until(lambda: script('return document.querySelector(".external-status")?.textContent.includes("Closed")'))
    for pid in (workspace / 'peer-pids').read_text().splitlines():
        try:
            os.kill(int(pid), 0)
        except ProcessLookupError:
            continue
        raise AssertionError('delegated peer was not reaped')
    script('const session=[...document.querySelectorAll("#sessions .session-row button")].find(e=>e.title===arguments[0]);if(!session)throw new Error("Original native Session is absent");session.click();return true', [native_session])
    until(lambda: script(r'return document.querySelector(".pane-session")?.title===arguments[0]&&!!document.querySelector("textarea[aria-label=\"Main message\"]")', [native_session]))
    (report / 'delegation.json').write_text(json.dumps({'same_conversation':True,'one_start':True,'reaped':True,'returned_native_session':native_session}))
